src/scenic/core/regions.py

```diff
@@ -281,883 +306,2816 @@ def __hash__(self):
+
+    def intersects(self, other, triedReversed=False):
+        if isinstance(other, PolygonalRegion):
+            if self.z != other.z:
+                return False
+
+        poly = toPolygon(other)
+        if poly is not None:
+            intersection = self.polygons & poly
+            return not intersection.is_empty
+        return super().intersects(other, triedReversed)
+
+    def intersect(self, other, triedReversed=False):
+        # If one of the regions isn't fixed fall back on default behavior
+        if needsSampling(self) or needsSampling(other):
+            return super().intersect(other, triedReversed)
+
+        if isinstance(other, PolygonalRegion):
+            if self.z != other.z:
+                return nowhere
+
+        poly = toPolygon(other)
+        orientation = other.orientation if self.orientation is None else self.orientation
+        if poly is not None:
+            intersection = self.polygons & poly
+            if isinstance(intersection, shapely.geometry.GeometryCollection):
+                if intersection.area > 0:
+                    poly_geoms = (shapely.geometry.Polygon, shapely.geometry.MultiPolygon)
+                    geoms = [geom for geom in intersection.geoms if isinstance(geom, poly_geoms)]
+                elif intersection.length > 0:
+                    line_geoms = (shapely.geometry.LineString, shapely.geometry.MultiLineString)
+                    geoms = [geom for geom in intersection.geoms if isinstance(geom, line_geoms)]
+                intersection = shapely.ops.unary_union(geoms)
+            orientation = orientationFor(self, other, triedReversed)
+            return regionFromShapelyObject(intersection, orientation=orientation)
+        return super().intersect(other, triedReversed)
+
+    def union(self, other, triedReversed=False, buf=0):
+        if isinstance(other, PolygonalRegion):
+            if self.z != other.z:
+                return super().union(other, triedReversed)
+
+        poly = toPolygon(other)
+
+        if not poly:
+            return super().union(other, triedReversed)
+        union = polygonUnion((self.polygons, poly), buf=buf)
+        orientation = VectorField.forUnionOf((self, other), tolerance=buf)
+        return PolygonalRegion(polygon=union, orientation=orientation)
+
+    def difference(self, other):
+        # If one of the regions isn't fixed fall back on default behavior
+        if needsSampling(self) or needsSampling(other):
+            return super().difference(other)
+
+        if isinstance(other, PolygonalRegion):
+            if self.z != other.z:
+                return self
+
+        poly = toPolygon(other)
+        if poly is not None:
+            diff = self.polygons - poly
+            return regionFromShapelyObject(diff, orientation=self.orientation)
+        return super().difference(other)
+
+    @staticmethod
+    def unionAll(regions, buf=0):
+        if not all([r.z == regions[0].z for r in regions]):
```

Comment:
Will fail if `regions` is a generator; better put back the original first line `regions = tuple(regions)`.